crates/agent/src/evolution.rs

```diff
@@ -455,35 +509,40 @@ impl ResourceSpecUpdater {
         Ok(ResourceSpecUpdater { pointers_by_image })
     }
 
-    /// Updates the given resource spec in place based on the x-collection-name.
+    /// Updates the given resource spec in place based on the x-collection-name annotation.
+    /// The new name
     fn update_resource_spec(
         &self,
         image_name: &str,
         materialization_name: &str,
-        new_collection_name: &str,
+        new_collection_name: Option<String>,
         resource_spec: &mut Value,
     ) -> anyhow::Result<()> {
         if let Some(pointer) = self.pointers_by_image.get(image_name) {
-            let has_existing = pointer
+            let Some(existing) = pointer
                 .query(&*resource_spec)
-                .map(|v| v.is_string())
-                .unwrap_or(false);
-            if has_existing {
-                let (_, new_base_name) = new_collection_name
-                    .rsplit_once('/')
-                    .expect("collection names must contain at least one forward slash");
-                if let Some(prev_val) = pointer.create_value(resource_spec) {
-                    let new_val = Value::String(new_base_name.to_owned());
-                    tracing::info!(%prev_val, %new_val, %materialization_name, %new_collection_name, "updating resource spec");
-                    *prev_val = new_val;
-                } else {
-                    anyhow::bail!("creating x-collection-name JSON location failed");
-                }
-            } else {
+                .and_then(|v| v.as_str()) else {
                 // Log noisily about this, but it's not clear that it shoudl be an error. It's
                 // possible, at least in principle, that a connector allows the field to be empty
                 // and will apply it's own default value based on the collection name.
-                tracing::warn!(%materialization_name, %new_collection_name, %pointer, "not updating resource spec because there is no existing value at that location");
+                tracing::warn!(%materialization_name, ?new_collection_name, %pointer, "not updating resource spec because there is no existing value at that location");
+                return Ok(())
+            };
+
+            let new_val = new_collection_name
+                .map(|n| {
+                    n.rsplit('/')
+                        .next()
+                        .expect("collection name must contain a slash")
+                        .to_owned()
+                })
+                .unwrap_or_else(|| next_name(existing).to_owned());
```

Comment:
We can get into trouble if there's a bump of _just_ the binding, followed later by a bump of the source collection, right?

Not saying we should do anything about it, just confirming my understanding.

One possible option would be to differentiate the way in which we increment (just) a resource config name, vs a collection name?

Reply:
Just documenting what we'd discussed on VC:
Yeah, this is weird. It seems like a more robust behavior would be to always only add the version suffix to the resource name, and ignore the new collection name completely. This doesn't seem like a change that's needed in this PR, but maybe something to consider for the future.